### PR TITLE
feat: upstream some nvim-treesitter functions

### PR DESCRIPTION
**Pull request by bfredl**
Rebase of a subset of #18232 . These are just the pure tree inspection functions, which I think we should merge already. The others are also useful but could require more review (how it interacts with editor state, LSP, etc, the best way).

### CODE REVIEW

File: runtime/lua/vim/treesitter/languagetree.lua [hunk 1]
```diff
@@ -549,6 +549,44 @@ function LanguageTree:contains(range)
...
+  for _, tree in pairs(self._trees) do
+    if tree_contains(tree, range) then
+      return tree
+    end
+  end
+
+  return nil
+end
+
+--- Gets the smallest named node that contains {range}
```
Comment (vigoux) [inline runtime/lua/vim/treesitter/languagetree.lua:580]:
```suggestion
--- Gets the smallest named node of this language that contains {range}
```

File: test/functional/treesitter/node_spec.lua [hunk 2]
```diff
@@ -59,4 +59,54 @@ describe('treesitter node API', function()
...
+        int x = 3;
+      }]])
+
+    exec_lua([[
+      tree = vim.treesitter.get_parser(0, "c"):parse()[1]
+      node = tree:root():child(0)
+      children = node:named_children()
+
+      len = 0
+      for _ in pairs(children) do len = len + 1 end
```
Comment (vigoux) [inline test/functional/treesitter/node_spec.lua:75]:
```suggestion
      return #children
```

File: runtime/lua/vim/treesitter/languagetree.lua (hunk 1 above)
Comment (bfredl) [reply runtime/lua/vim/treesitter/languagetree.lua:580]: hmm, is this right? this gets the smallest node of any language in the LanguageTree?

**Comment by clason** [comment]
Documentation nit: parameter annotations shouldn't start with `The` or `A` -- just remove these. (Not a blocker for this PR as the whole `vim.treesitter` documentation needs bigger cleaning up; I just thought I'd mention it so the pattern doesn't get repeated mindlessly later.)

**Review by justinmk — APPROVED** [review]

**Pull request merged**